hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDTableServiceClient.java

```diff
@@ -36,17 +39,58 @@
 
 public class SparkRDDTableServiceClient<T> extends BaseHoodieTableServiceClient<HoodieData<HoodieRecord<T>>, HoodieData<WriteStatus>, JavaRDD<WriteStatus>> {
+
+  private HoodieMetadataWriteWrapper metadataWriterWrapper = new HoodieMetadataWriteWrapper();
   protected SparkRDDTableServiceClient(HoodieEngineContext context,
                                        HoodieWriteConfig clientConfig,
                                        Option<EmbeddedTimelineService> timelineService) {
     super(context, clientConfig, timelineService);
   }
 
   @Override
-  protected List<HoodieWriteStat> triggerWritesAndFetchWriteStats(HoodieWriteMetadata<JavaRDD<WriteStatus>> writeMetadata) {
-    return writeMetadata.getWriteStatuses().map(writeStatus -> writeStatus.getStat()).collect();
+  protected Pair<List<HoodieWriteStat>, List<HoodieWriteStat>> triggerWritesAndFetchWriteStats(HoodieWriteMetadata<JavaRDD<WriteStatus>> tableServiceWriteMetadata) {
+    // Triggering the dag for writes.
+    // If streaming writes are enabled, writes to both data table and metadata table gets triggered at this juncture.
+    // If not, writes to data table gets triggered here.
+    // When streaming writes are enabled, data table's WriteStatus is expected to contain all stats required to generate metadata table records and so each object will be larger.
+    // So, here we are dropping all additional stats and error records to retain only the required information and prevent collecting large objects on the driver.
+    List<SparkRDDWriteClient.SlimWriteStats> writeStatusMetadataTrackerList = tableServiceWriteMetadata.getWriteStatuses()
+        .map(writeStatus -> new SparkRDDWriteClient.SlimWriteStats(writeStatus.isMetadataTable(), writeStatus.getTotalRecords(), writeStatus.getTotalErrorRecords(),
+            writeStatus.getStat())).collect();
+
+    List<HoodieWriteStat> dataTableWriteStats = writeStatusMetadataTrackerList.stream().filter(entry -> !entry.isMetadataTable()).map(entry -> entry.getWriteStat()).collect(Collectors.toList());
+    List<HoodieWriteStat> mdtWriteStats = writeStatusMetadataTrackerList.stream().filter(entry -> entry.isMetadataTable()).map(entry -> entry.getWriteStat()).collect(Collectors.toList());
+
+    if (isMetadataTable) {
+      ValidationUtils.checkArgument(dataTableWriteStats.isEmpty(), "For Metadata table,"
+          + "we do not expect any writes having WriteStatus referring to data table. ");
+      dataTableWriteStats.clear();
+      dataTableWriteStats.addAll(mdtWriteStats);
+      mdtWriteStats.clear();
+    }
+    return Pair.of(dataTableWriteStats, mdtWriteStats);
+  }
+
+  @Override
+  protected HoodieWriteMetadata<HoodieData<WriteStatus>> processWriteMetadata(HoodieTable table, HoodieWriteMetadata<HoodieData<WriteStatus>> writeMetadata, String instantTime) {
+    if (!isMetadataTable && config.isMetadataTableEnabled() && config.isMetadataStreamingWritesEnabled(table.getMetaClient().getTableConfig().getTableVersion())) {
```

Comment:
This condition is the same as line 87 and I think there are similar checks in the SparkRDDWriteClient, is there a common place to put this logic to keep it consistent?

Reply:
yes, there are minor difference in conditions we check in SparkRDDWriteClient and in the table service client. 
but I have tried to unify common denominator.